### PR TITLE
Full texture cache with palette updating.

### PR DESCRIPTION
**Pull request by Interrupt**
This implements a more fully featured texture cache for all bitmaps, not just walls. It also does an update-in-place of textures once per frame in case the palette changed so that palette animations work again.

### CODE REVIEW

File: src/GameSrc/cybmem.c [hunk 1]
```diff
@@ -40,6 +40,7 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
 #include "Shock.h"
 #include "sideicon.h"
 #include "criterr.h"
+#include "OpenGl.h"
```
Comment (inguin) [inline src/GameSrc/cybmem.c:43]: Wrong case, must be `OpenGL.h`.

File: src/MacSrc/OpenGL.cc [hunk 2]
```diff
@@ -292,15 +345,23 @@ static void convert_texture(GLuint texture, grs_bitmap *bm) {
...
 
-    glBindTexture(GL_TEXTURE_2D, texture);
-    glTexImage2D(GL_TEXTURE_2D, 0, GL_RGBA, bm->w, bm->h, 0, GL_RGBA, GL_UNSIGNED_BYTE, rgba->pixels);
-
-    SDL_FreeSurface(rgba);
-    SDL_FreeSurface(surface);
+    // Cache this new surface.
+    CachedTexture ct;
+    ct.bitmap = *surface;
+    ct.converted = *rgba;
```
Comment (inguin) [inline src/MacSrc/OpenGL.cc]: This does not work! You're copying dynamically allocated `SDL_Surface` structures here. The allocated memory will leak, and `opengl_clear_texture_cache()` will segfault when calling `SDL_FreeSurface()`. You must store the original pointers.

File: src/MacSrc/OpenGL.cc [hunk 3]
```diff
@@ -292,15 +345,23 @@ static void convert_texture(GLuint texture, grs_bitmap *bm) {
...
+
+    bool cached = opengl_cache_texture(ct, bm);
+    if(!cached) {
+        DEBUG("Not enough room to cache texture!");
+        SDL_FreeSurface(surface);
+        SDL_FreeSurface(rgba);
+    }
 }
 
 void opengl_cache_texture(int idx, int size, grs_bitmap *bm) {
```
Comment (inguin) [inline src/MacSrc/OpenGL.cc:374]: This method is now useless. It's still called by `load_textures()` during program initialization, but the cache will then be cleared when starting a new game.

File: src/MacSrc/OpenGL.cc [hunk 4]
```diff
@@ -283,7 +291,52 @@ void opengl_set_viewport(int x, int y, int width, int height) {
...
+        return;
+
+    DEBUG("Clearing OpenGL texture cache.");
+
+    std::map<uint8_t *, CachedTexture>::iterator iter;
+    for(iter = texturesByBitsPtr.begin(); iter != texturesByBitsPtr.end(); iter++) {
+        CachedTexture ct;
+        SDL_FreeSurface(&ct.bitmap);
+        SDL_FreeSurface(&ct.converted);
+        glDeleteTextures(1, &ct.texture);
```
Comment (inguin) [inline src/MacSrc/OpenGL.cc]: `ct` is uninitialized here. Should be: `CachedTexture& ct = iter->second;`.

File: src/MacSrc/OpenGL.cc [hunk 5]
```diff
@@ -283,7 +291,52 @@ void opengl_set_viewport(int x, int y, int width, int height) {
     glViewport(scaled_x, scaled_y, scaled_width, scaled_height);
 }
 
-static void convert_texture(GLuint texture, grs_bitmap *bm) {
+bool opengl_cache_texture(CachedTexture toCache, grs_bitmap *bm) {
```
Comment (inguin) [inline src/MacSrc/OpenGL.cc]: I recommend making purely local functions that are not used from anywhere else `static`. It guarantees that nobody else is using them and allows the compiler to optimize (inline).

**Comment by inguin** [comment]
Found some more issues:

- On my machine, this is significantly slower than before because `glTexImage2D()` is now performed for every surface drawn. It should be enough to do this once per frame and texture.
- After a few minutes of playing the engine stopped with "Memory error: Not enough cache memory to load bitmap!". Looks like a bad memleak somewhere.
- Text overlays on walls are often broken, after playing for a while they all read the same. I ran into the same issue when implementing my own texture cache; I suspect the same bitmap pointer is reused for different texts.

**Comment by Interrupt** [comment]
Fixed a bunch of issues, thanks for the good review! Should be much faster now that world textures only get updated once per frame, and we could do that even less if we kept track of when a palette change happened. Is this any faster for you now?

**Comment by Interrupt** [comment]
Fixed a bunch of issues, thanks for the good review! Should be much faster now that world textures only get updated once per frame, and we could do that even less if we kept track of when a palette change happened. Is this any faster for you now?

**Comment by inguin** [comment]
Yes, that's much faster again! The game now uses about 50% CPU (one core), but that seems to be a baseline. The software renderer at minimum resolution uses as much.

There's still the compile error because of the wrong case of the include name in `cybmem.c`, but the remaining issues are fixed. Oh, and it disabled my nice hack to always use the maximum-resolution version of textures, but I can restore that after the merge.

**Comment by Interrupt** [comment]
The hack to use the max res version of textures never was working for me, do we need to go disable that feature where it's actually happening in opengl mode?

**Comment by inguin** [comment]
That's weird, how could that not have worked? It was super-simple: In `texturesByBitsPtr`, the entries for all texture sizes were pointing to the same GL texture.

**Comment by inguin** [comment]
I spent some more time thinking about this. After the latest changes, what is the benefit of caching non-locked surfaces? They will be fully updated (RSD8 conversion, Palette->RGBA, glTexImage2D) for every drawn surface. I believe this solution is now equivalent to my much simpler cache, except that the latter lacked palette updating.

**Pull request merged**